# Collecting and Selecting Data

In this section, we are going to dive into the basic collecting and cleaning the dataset. This will bring us to load a dataset, assess its features, and glean as much context as possible. We will then use this to narrow and clean our data to something of interest. 

A common mistake data science professionals make is diving straight into a dataset and forgetting to ask basic questions, like where did the dataset come from? I will argue that bias in data is inevitable, because reality itself is biased. We will then use this to mindfully think through how this data was collected and base decisions accordingly on it. 

## The Bird Strike Dataset 

Aircraft bird strikes as reported by the Federal Aviation Administration (FAA) in the United States. A bird strike occurs when a bird collides with an aircraft, and the damage can be severe. Each year, there are on average 13,000 birdstrikes in the United States alone and cost the aviation industry $400 million in damages. While most bird strikes are minor, some can be dangerous and fatal. 

<a title="Greg L, CC BY 2.0 &lt;https://creativecommons.org/licenses/by/2.0&gt;, via Wikimedia Commons" href="https://commons.wikimedia.org/wiki/File:US_Airways_Flight_1549_(N106US)_after_crashing_into_the_Hudson_River_(crop_2).jpg"><img width="512" alt="US Airways Flight 1549 (N106US) after crashing into the Hudson River (crop 2)" src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/8f/US_Airways_Flight_1549_%28N106US%29_after_crashing_into_the_Hudson_River_%28crop_2%29.jpg/512px-US_Airways_Flight_1549_%28N106US%29_after_crashing_into_the_Hudson_River_%28crop_2%29.jpg?20200816213116"></a>

**In 2009, US Airways Flight 1549 suffered a major birdstrike resulting in an emergency landing in the Hudson River.**<br>
*Greg L, CC BY 2.0, via Wikimedia Commons*

To get started use `pip` or `conda` to install `openpyxl`. Pandas will  need this library to load Excel files. If you are using plain Python, make sure to install `pandas` too. 

In [1]:
!pip install openpyxl pandas

In [2]:
!conda install openpyxl pandas --yes

zsh:1: command not found: conda


## FAA Birdstrike Dataset 

We are going to use the FAA bird strike dataset for this data exploration exercise. This dataset was obtained from the [FAA Wildlife Strike Database](https://wildlife.faa.gov/search). To keep the experience raw, I have not cleaned or converted the exported Excel `xlsx` file into a different format like SQLite, CSV, or Pickle. 

Below, let's load the Excel file into a Pandas `DataFrame`. This file is about 129MB so it may take a moment. 

In [3]:
import pandas as pd

df = pd.read_excel('bird_strike_faa.xlsx')

Let's display the DataFrame with all of its columns. 

In [4]:
with pd.option_context('display.max_columns', None):
  display(df)

INDEX_NR INCIDENT_DATE  INCIDENT_MONTH  INCIDENT_YEAR   TIME  \
0         608242    1996-06-22               6           1996    NaN   
1         608243    1996-06-26               6           1996    NaN   
2         608244    1996-07-01               7           1996    NaN   
3         608245    1996-07-01               7           1996    NaN   
4         608246    1996-07-01               7           1996    NaN   
...          ...           ...             ...            ...    ...   
298196   1516465    2024-03-17               3           2024  07:15   
298197   1516467    2024-03-17               3           2024  19:15   
298198   1516468    2024-03-17               3           2024  16:39   
298199   1516469    2024-03-17               3           2024  21:45   
298200   1516470    2024-03-17               3           2024  20:00   

       TIME_OF_DAY AIRPORT_ID                      AIRPORT   LATITUDE  \
0              NaN       KSMF              SACRAMENTO INTL   38.69542   
1              NaN       KDEN          DENVER INTL AIRPORT   39.85841   
2              NaN       KOMA              EPPLEY AIRFIELD   41.30252   
3              NaN       KIAD  WASHINGTON DULLES INTL ARPT   38.94453   
4              NaN       KLGA              LA GUARDIA ARPT   40.77724   
...            ...        ...                          ...        ...   
298196         NaN       KSEA          SEATTLE-TACOMA INTL   47.44898   
298197         NaN       KHYI    SAN MARCOS MUNICIPAL ARPT   29.89361   
298198         Day       KSBA      SANTA BARBARA MUNICIPAL   34.42621   
298199       Night       ZGGG               BAIYUN AIRPORT  23.392436   
298200         NaN       ZZZZ                      UNKNOWN        NaN   

         LONGITUDE RUNWAY STATE FAAREGION       LOCATION OPID  \
0       -121.59077    NaN    CA      AWP             NaN  UAL   
1         -104.667    NaN    CO      ANM             NaN  UAL   
2        -95.89417    NaN    NE      ACE             NaN  UAL   
3        -77.45581    NaN    DC      AEA             NaN  UAL   
4        -73.87261    NaN    NY      AEA             NaN  UAL   
...            ...    ...   ...       ...            ...  ...   
298196  -122.30931    16L    WA      ANM             NaN  UNK   
298197   -97.86469      8    TX      ASW             NaN  EJA   
298198  -119.84037    NaN    CA      AWP             NaN  ASA   
298199  113.298786    20L    FN      FGN             NaN  FDX   
298200         NaN    NaN   NaN       NaN  FOUND AT KABE  ATN   

                  OPERATOR     REG   FLT    AIRCRAFT  AMA  AMO   EMA   EMO  \
0          UNITED AIRLINES     NaN  1768   B-737-300  148   24  10.0   1.0   
1          UNITED AIRLINES     NaN  1845   B-737-300  148   24  10.0   1.0   
2          UNITED AIRLINES     NaN   306   B-757-200  148   26  34.0  40.0   
3          UNITED AIRLINES     NaN   510       A-320  04A    3  23.0   1.0   
4          UNITED AIRLINES     NaN   677       A-320  04A    3  23.0   1.0   
...                    ...     ...   ...         ...  ...  ...   ...   ...   
298196             UNKNOWN     NaN   NaN     UNKNOWN  NaN  NaN   NaN   NaN   
298197             NETJETS  N237QS   237  CL-601/604  188    7  22.0   4.0   
298198     ALASKA AIRLINES  N590AS  1141   B-737-800  148   43  10.0   1.0   
298199       FEDEX EXPRESS  N103FE  6024   B-767-300  148   97  22.0   7.0   
298200  AIR TRANSPORT INTL  N1439A  3495   B-767-300  148   97  22.0   7.0   

       AC_CLASS  AC_MASS TYPE_ENG  NUM_ENGS  ENG_1_POS  ENG_2_POS  ENG_3_POS  \
0           A        4.0        D       2.0        1.0        1.0        NaN   
1           A        4.0        D       2.0        1.0        1.0        NaN   
2           A        4.0        D       2.0        1.0        1.0        NaN   
3           A        4.0        D       2.0        1.0        1.0        NaN   
4           A        4.0        D       2.0        1.0        1.0        NaN   
...         ...      ...      ...       ...        ...        ...   

Yikes. Those are a lot of columns. Just how many rows and columns are we dealing with? We can access this by running `df.shape`. 

In [5]:
df.shape

(298201, 101)

So we got 298,201 rows and 101 columns. We will thoughtfully go through these and narrow our dataset only to certain columns and rows. We will save this work by saving it for later. To get there though, let's try to understand our data first.

We have another file that might be useful: the `read_me.xlsx`. Included are sheets in an Excel workbook that have descriptions for each column as well as engine codes, aircraft types, and engine position. Those last three sheets are helpful to make sense of certain columns that use coding conventions. Let's display all the column description in a `DataFrame` and make sure no truncation happens setting `display.max_rows` and `display.max_colwidth` to `None`. 

In [6]:
schema = pd.read_excel('read_me.xlsx',sheet_name='Column Name')

with pd.option_context('display.max_rows', None, 'display.max_colwidth', None):
    display(schema)

Column Name  \
0                   INDEX NR   
1                       OPID   
2                   OPERATOR   
3                      ATYPE   
4                        AMA   
5                        AMO   
6                        EMA   
7                        EMO   
8                   AC_CLASS   
9                    AC_MASS   
10                  NUM_ENGS   
11                  TYPE_ENG   
12                 ENG_1_POS   
13                 ENG_2_POS   
14                 ENG_3_POS   
15                 ENG_4_POS   
16                       REG   
17                       FLT   
18         REMAINS_COLLECTED   
19              REMAINS_SENT   
20             INCIDENT_DATE   
21            INCIDENT_MONTH   
22             INCIDENT_YEAR   
23               TIME_OF_DAY   
24                      TIME   
25                AIRPORT_ID   
26                   AIRPORT   
27                     STATE   
28                 FAAREGION   
29                   ENROUTE   
30                    RUNWAY   
31                  LOCATION   
32                    HEIGHT   
33                     SPEED   
34                  DISTANCE   
35              PHASE_OF_FLT   
36                    DAMAGE   
37                     Blank   
38                  N = None   
39                 M = Minor   
40   M? = Undetermined level   
41           S = Substantial   
42             D = Destroyed   
43                   STR_RAD   
44                   DAM_RAD   
45              STR_WINDSHLD   
46              DAM_WINDSHLD   
47                  STR_NOSE   
48                  DAM_NOSE   
49                  STR_ENG1   
50                  DAM_ENG1   
51                  ING_ENG1   
52                  STR_ENG2   
53                  DAM_ENG2   
54                  ING_ENG2   
55                  STR_ENG3   
56                  DAM_ENG3   
57                  ING_ENG3   
58                  STR_ENG4   
59                  DAM_ENG4   
60                  ING_ENG4   
61            INGESTED_OTHER   
62                  STR_PROP   
63                  DAM_PROP   
64              STR_WING_ROT   
65              DAM_WING_ROT   
66                  STR_FUSE   
67                  DAM_FUSE   
68                    STR_LG   
69                    DAM_LG   
70                  STR_TAIL   
71                  DAM_TAIL   
72                 STR_LGHTS   
73                 DAM_LGHTS   
74                 STR_OTHER   
75                 DAM_OTHER   
76             OTHER_SPECIFY   
77                    EFFECT   
78              EFFECT_OTHER   
79                       SKY   
80                    PRECIP   
81          BIRD_BAND_NUMBER   
82                SPECIES_ID   
83                   SPECIES   
84                BIRDS_SEEN   
85              BIRDS_STRUCK   
86                      SIZE   
87                    WARNED   
88                  COMMENTS   
89                   REMARKS   
90                       AOS   
91              COST_REPAIRS   
92                COST_OTHER   
93     COST_REPAIRS_INFL_ADJ   
94       COST_OTHER_INFL_ADJ   
95             REPORTED_NAME   
96            REPORTED_TITLE   
97             REPORTED_DATE   
98                    SOURCE   
99                    PERSON   
100              NR_INJURIES   
101            NR_FATALITIES   
102                  LUPDATE   
103                 TRANSFER   
104         INDICATED_DAMAGE   

                                                                                                                                                                                                                                                                                                                                                                                                                              Explanation of Column Name and Codes  
0                                                                                                                                                                          

Oh man... talk about drinking from a firehose! But at least we have a descriptor of every column. Let's take a look at every data type of every column next. 

In [7]:
with pd.option_context('display.max_rows', None):
    display(df.dtypes)

INDEX_NR                   int64
INCIDENT_DATE             object
INCIDENT_MONTH             int64
INCIDENT_YEAR              int64
TIME                      object
TIME_OF_DAY               object
AIRPORT_ID                object
AIRPORT                   object
LATITUDE                  object
LONGITUDE                 object
RUNWAY                    object
STATE                     object
FAAREGION                 object
LOCATION                  object
OPID                      object
OPERATOR                  object
REG                       object
FLT                       object
AIRCRAFT                  object
AMA                       object
AMO                       object
EMA                      float64
EMO                      float64
AC_CLASS                  object
AC_MASS                  float64
TYPE_ENG                  object
NUM_ENGS                 float64
ENG_1_POS                float64
ENG_2_POS                float64
ENG_3_POS                float64
ENG_4_POS 

This is not something we just want to gloss over. We really want to take time to understand this data, ask questions, and consult experts if needed. It is important to understand what we are trying to achieve here: find possible causes for bird strikes. 

**PLEASE ASK QUESTIONS! DON'T WORK WITH DATA YOU DON'T FULLY UNDERSTAND. EVER!**

## Some Basic Cleaning

Based on these column descriptors, it is safe to set `INDEX_NR` to be our `DataFrame` index. Let's get that done. This is the unique identifier for each record.  

In [8]:
df.set_index('INDEX_NR', inplace=True)

Let's say we talked to a weather-hardened pilot to give us further context to these descriptions. We also found a friendly retired expert from the FAA who knows this dataset really well. Our curiosity starts to rewards us and we begin to understand the bigger picture of this dataset. We also take the opportunity to ask them their opinions on what causes bird strikes. Perhaps we are biasing our future conclusions already, but it is always a good idea to consult the experts on everything you can once you have them. 

![](resource/ZPGDGucW.svg)

Because of this knowledge, we start to have a good idea of what data we want to include. For starts, we want to only include recent data from 2015 onwards. We have deemed there is little value looking at older data from before that year. We can argue that the nature of bird strikes never change, but the environment certaintly changes. Schedules and airports grow and shrink, weather patterns change, and different airlines come and go.  Even the [FAA themselves](https://wildlife.faa.gov/home) say:  

> Expanding wildlife populations, increases in number of aircraft movements, a trend toward faster and quieter aircraft, and outreach to the aviation community all have contributed to the observed increase in reported wildlife strikes.

So nothing is ever static and things are always changing. For this reason, we make a judgment to limit our data to 2015 onwards. Let's delegate this to Pandas. 

In [9]:
df = df[df['INCIDENT_YEAR'] >= 2015]

For the record, we can use `>=`, `<=`, `<`, `>`, and `==` for equality and inequality filtering. 

After removing everything before 2015, how many rows do we have now? 

In [10]:
df.shape

(141069, 100)

So we went from 298,201 to 141,069 rows. This substantially removed a lot of data which will make working with the data easier. We will have less data to sift through. Just be mindful and document every decision you make along the way. Every decision is a bias, and while bias inherently is not bad, it will steer your conclusions whether they are right or wrong.  

Let's take a look at our dataframe but alphabetically sort the columns to make them easy to pick through

In [11]:
with pd.option_context('display.max_columns', None):
    display(df.sort_index(axis = 1))

AC_CLASS  AC_MASS      AIRCRAFT                          AIRPORT  \
INDEX_NR                                                                    
708307        A        1.0         PA-28             VERO BEACH MUNICIPAL   
708308        A        3.0       BE-1900             KENAI MUNICIPAL ARPT   
708309        A        1.0  PA-46 MALIBU  DAVID WAYNE HOOKS MEMORIAL ARPT   
708310        A        4.0     B-717-200            LAMBERT-ST LOUIS INTL   
708311        A        2.0    BE-90 KING            POMPANO BEACH AIRPARK   
...           ...      ...           ...                              ...   
1516465       NaN      NaN       UNKNOWN              SEATTLE-TACOMA INTL   
1516467       A        3.0    CL-601/604        SAN MARCOS MUNICIPAL ARPT   
1516468       A        4.0     B-737-800          SANTA BARBARA MUNICIPAL   
1516469       A        4.0     B-767-300                   BAIYUN AIRPORT   
1516470       A        4.0     B-767-300                          UNKNOWN   

         AIRPORT_ID  AMA  AMO   AOS BIRD_BAND_NUMBER            COMMENTS  \
INDEX_NR                                                                   
708307         KVRB  729   23   NaN              NaN   /Legacy Record=1/   
708308         PAEN  123   27   NaN              NaN   /Legacy Record=2/   
708309         KDWH  729   45   NaN              NaN   /Legacy Record=4/   
708310         KSTL  148   45   NaN              NaN   /Legacy Record=6/   
708311         KPMP  NaN  NaN   NaN              NaN   /Legacy Record=7/   
...             ...  ...  ...   ...              ...                 ...   
1516465        KSEA  NaN  NaN   NaN              NaN                 NaN   
1516467        KHYI  188    7   NaN              NaN                 NaN   
1516468        KSBA  148   43  0.35              NaN                 NaN   
1516469        ZGGG  148   97   NaN              NaN                 NaN   
1516470        ZZZZ  148   97   NaN              NaN                 NaN   

          COST_OTHER  COST_OTHER_INFL_ADJ  COST_REPAIRS  \
INDEX_NR                                                  
708307           NaN                  NaN           NaN   
708308           NaN                  NaN           NaN   
708309           NaN                  NaN           NaN   
708310           NaN                  NaN           NaN   
708311           NaN                  NaN           NaN   
...              ...                  ...           ...   
1516465          NaN                  NaN           NaN   
1516467          NaN                  NaN           NaN   
1516468          NaN                  NaN           NaN   
1516469          NaN                  NaN           NaN   
1516470          NaN                  NaN           NaN   

          COST_REPAIRS_INFL_ADJ DAMAGE_LEVEL  DAM_ENG1  DAM_ENG2  DAM_ENG3  \
INDEX_NR                                                                     
708307                      NaN            M         0         0         0   
708308                      NaN            M         0         0         0   
708309                      NaN            M         1         0         0   
708310                      NaN            M         0         0         0   
708311                      NaN            M         0         0         0   
...                         ...          ...       ...       ...       ...   
1516465                     NaN          NaN         0         0         0   
1516467                     NaN            N         0         0         0   
1516468                     NaN            N         0         0         0   
1516469                     NaN            N         0         0         0   
1516470                     NaN          NaN         0         0         0   

          DAM_ENG4  DAM_FUSE  DAM_LG  DAM_LGHTS  DAM_NOSE  DAM_OTHER  \
INDEX_NR                                                               
708307           0         0       0          0         0          0   
708308           0 

Now let's weigh if we want to omit any columns. Granted we are going to pick and choose columns to aid us in making decisions and further clean our data in the next section. But we should consider what columns we want to immediately exclude that just add noise. 

In [12]:
chosen_cols = ["OPID",
"OPERATOR",
"AIRCRAFT",
"AC_CLASS",
"AC_MASS",
"NUM_ENGS",
"TYPE_ENG",
"INCIDENT_DATE",
"INCIDENT_YEAR", 
"INCIDENT_MONTH",
"TIME_OF_DAY",
"TIME",
"AIRPORT_ID",
"AIRPORT",
"STATE",
"RUNWAY",
"LOCATION",
"LATITUDE",
"LONGITUDE",
"HEIGHT",
"SPEED",
"DISTANCE",
"PHASE_OF_FLIGHT",
"DAMAGE_LEVEL",
"STR_RAD",
"DAM_RAD",
"STR_WINDSHLD",
"DAM_WINDSHLD",
"STR_NOSE",
"STR_PROP",
"DAM_PROP",
"STR_WING_ROT",
"DAM_WING_ROT",
"STR_FUSE",
"DAM_FUSE",
"STR_LG",
"DAM_LG",
"STR_TAIL",
"DAM_TAIL",
"STR_LGHTS",
"DAM_LGHTS",
"STR_OTHER",
"EFFECT",
"SKY",
"PRECIPITATION",
"SPECIES_ID",
"SPECIES",
"SIZE",
"WARNED",
"COST_REPAIRS",
"COST_OTHER",
"COST_REPAIRS_INFL_ADJ",
"COST_OTHER_INFL_ADJ",
"NR_INJURIES",
"NR_FATALITIES",
"INDICATED_DAMAGE"]

df = df[chosen_cols]

with pd.option_context('display.max_columns', None):
    display(df)

OPID            OPERATOR      AIRCRAFT AC_CLASS  AC_MASS  NUM_ENGS  \
INDEX_NR                                                                      
708307    BUS            BUSINESS         PA-28      A        1.0       1.0   
708308    BUS            BUSINESS       BE-1900      A        3.0       2.0   
708309    BUS            BUSINESS  PA-46 MALIBU      A        1.0       1.0   
708310    DAL     DELTA AIR LINES     B-717-200      A        4.0       2.0   
708311    BUS            BUSINESS    BE-90 KING      A        2.0       2.0   
...       ...                 ...           ...      ...      ...       ...   
1516465   UNK             UNKNOWN       UNKNOWN      NaN      NaN       NaN   
1516467   EJA             NETJETS    CL-601/604      A        3.0       2.0   
1516468   ASA     ALASKA AIRLINES     B-737-800      A        4.0       2.0   
1516469   FDX       FEDEX EXPRESS     B-767-300      A        4.0       2.0   
1516470   ATN  AIR TRANSPORT INTL     B-767-300      A        4.0       2.0   

         TYPE_ENG INCIDENT_DATE  INCIDENT_YEAR  INCIDENT_MONTH TIME_OF_DAY  \
INDEX_NR                                                                     
708307          A    2015-05-22           2015               5         NaN   
708308          C    2015-06-18           2015               6         NaN   
708309          A    2015-09-20           2015               9         NaN   
708310          D    2015-11-07           2015              11         NaN   
708311          C    2015-12-17           2015              12         NaN   
...           ...           ...            ...             ...         ...   
1516465       NaN    2024-03-17           2024               3         NaN   
1516467         D    2024-03-17           2024               3         NaN   
1516468         D    2024-03-17           2024               3         Day   
1516469         D    2024-03-17           2024               3       Night   
1516470         D    2024-03-17           2024               3         NaN   

           TIME AIRPORT_ID                          AIRPORT STATE RUNWAY  \
INDEX_NR                                                                   
708307      NaN       KVRB             VERO BEACH MUNICIPAL    FL      4   
708308      NaN       PAEN             KENAI MUNICIPAL ARPT    AK    NaN   
708309      NaN       KDWH  DAVID WAYNE HOOKS MEMORIAL ARPT    TX    NaN   
708310      NaN       KSTL            LAMBERT-ST LOUIS INTL    MO    30R   
708311      NaN       KPMP            POMPANO BEACH AIRPARK    FL     15   
...         ...        ...                              ...   ...    ...   
1516465   07:15       KSEA              SEATTLE-TACOMA INTL    WA    16L   
1516467   19:15       KHYI        SAN MARCOS MUNICIPAL ARPT    TX      8   
1516468   16:39       KSBA          SANTA BARBARA MUNICIPAL    CA    NaN   
1516469   21:45       ZGGG                   BAIYUN AIRPORT    FN    20L   
1516470   20:00       ZZZZ                          UNKNOWN   NaN    NaN   

               LOCATION   LATITUDE   LONGITUDE  HEIGHT  SPEED  DISTANCE  \
INDEX_NR                                                                  
708307              NaN   27.65556   -80.41794     NaN    NaN       NaN   
708308              NaN     60.572  -151.24753     NaN    NaN       NaN   
708309              NaN   30.06186   -95.55278     NaN    NaN       NaN   
708310              NaN   38.74769   -90.35999     NaN    NaN       2.0   
708311              NaN   26.24714   -80.11106     0.0    NaN       0.0   
...                 ...        ...         ...     ...    ...       ...   
1516465             NaN   47.44898  -122.30931     NaN    NaN       0.0   
1516467             NaN   29.89361   -97.86469  2000.0  160.0       5.0   
1516468             NaN   34.42621  -119.84037    30.0    NaN       0.0   
1516469             NaN  23.392436  113.298786  3000.0  180.0      10.0   
1516470   FOUND AT KABE        NaN         NaN     NaN    NaN       NaN   

 

## Saving Our Data

To save our data, let's just save to a plain CSV using `to_csv()` in Pandas. Make sure to also give the indexed column a name, which we will just . 

In [13]:
df.to_csv('birdstrike_section1.csv', index_label='INDEX_NR')

For good measure, let's make sure we can read the CSV without issue into Pandas. 

In [14]:
pd.read_csv('birdstrike_section1.csv', index_col='INDEX_NR')

/var/folders/r5/1fmf39d96332_2dgz7yjrw5w0000gn/T/ipykernel_14831/967290845.py:1: DtypeWarning: Columns (18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('birdstrike_section1.csv', index_col='INDEX_NR')


OPID            OPERATOR      AIRCRAFT AC_CLASS  AC_MASS  NUM_ENGS  \
INDEX_NR                                                                      
708307    BUS            BUSINESS         PA-28      A        1.0       1.0   
708308    BUS            BUSINESS       BE-1900      A        3.0       2.0   
708309    BUS            BUSINESS  PA-46 MALIBU      A        1.0       1.0   
708310    DAL     DELTA AIR LINES     B-717-200      A        4.0       2.0   
708311    BUS            BUSINESS    BE-90 KING      A        2.0       2.0   
...       ...                 ...           ...      ...      ...       ...   
1516465   UNK             UNKNOWN       UNKNOWN      NaN      NaN       NaN   
1516467   EJA             NETJETS    CL-601/604      A        3.0       2.0   
1516468   ASA     ALASKA AIRLINES     B-737-800      A        4.0       2.0   
1516469   FDX       FEDEX EXPRESS     B-767-300      A        4.0       2.0   
1516470   ATN  AIR TRANSPORT INTL     B-767-300      A        4.0       2.0   

         TYPE_ENG INCIDENT_DATE  INCIDENT_YEAR  INCIDENT_MONTH  ...  \
INDEX_NR                                                        ...   
708307          A    2015-05-22           2015               5  ...   
708308          C    2015-06-18           2015               6  ...   
708309          A    2015-09-20           2015               9  ...   
708310          D    2015-11-07           2015              11  ...   
708311          C    2015-12-17           2015              12  ...   
...           ...           ...            ...             ...  ...   
1516465       NaN    2024-03-17           2024               3  ...   
1516467         D    2024-03-17           2024               3  ...   
1516468         D    2024-03-17           2024               3  ...   
1516469         D    2024-03-17           2024               3  ...   
1516470         D    2024-03-17           2024               3  ...   

                     SPECIES   SIZE   WARNED COST_REPAIRS COST_OTHER  \
INDEX_NR                                                               
708307          Unknown bird    NaN  Unknown          NaN        NaN   
708308          Unknown bird    NaN  Unknown          NaN        NaN   
708309          Unknown bird    NaN  Unknown          NaN        NaN   
708310          Unknown bird    NaN  Unknown          NaN        NaN   
708311          Unknown bird    NaN  Unknown          NaN        NaN   
...                      ...    ...      ...          ...        ...   
1516465       American robin  Small  Unknown          NaN        NaN   
1516467          Meadowlarks  Small       No          NaN        NaN   
1516468   Western meadowlark  Small  Unknown          NaN        NaN   
1516469         Unknown bird    NaN  Unknown          NaN        NaN   
1516470         Unknown bird    NaN  Unknown          NaN        NaN   

         COST_REPAIRS_INFL_ADJ COST_OTHER_INFL_ADJ NR_INJURIES NR_FATALITIES  \
INDEX_NR                                                                       
708307                     NaN                 NaN         NaN           NaN   
708308                     NaN                 NaN         NaN           NaN   
708309                     NaN                 NaN         NaN           NaN   
708310                     NaN                 NaN         NaN           NaN   
708311                     NaN                 NaN         NaN           NaN   
...                        ...                 ...         ...           ...   
1516465                    NaN                 NaN         NaN           NaN   
1516467                    NaN                 NaN         NaN           NaN   
1516468                    NaN                 NaN         NaN           NaN   
1516469                    NaN                 NaN         NaN           NaN   
1516470                    NaN                 NaN         NaN           NaN   

          INDICATED_DAMAGE  
INDEX_NR                    
708307               

Great! No issues here. We have successfully narrowed our data down to what we think will be of interest. Again remember to have an auditable trail of how we transformed and used this data. Keep the original data handy in case we ever need additional rows or columns again. But this should give us plenty to work with going forward. 

Note we could have also saved the results to a pickle using `to_pickle()` or SQL using `to_sql()`. I personally like using SQLite to store data [and offer an entire Anaconda course on that topic](https://learning.anaconda.cloud/introduction-to-sql). 

## Where Did the Data Come From? 

Before we wrap up this section, let's think carefully and step back. We talked to an experienced pilot and retired FAA official. They were wonderful and shared some great information on this data. But being data experts, we have to truly ask what created this data. 

<a title="See page for author, Public domain, via Wikimedia Commons" href="https://commons.wikimedia.org/wiki/File:F16_after_bird_strike.jpg"><img width="512" alt="F16 after bird strike" src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/de/F16_after_bird_strike.jpg/512px-F16_after_bird_strike.jpg?20050925122118"></a>

**Bird strike on an F-16 canopy**

Each record is reported by a pilot to the FAA. Now stop and consider that. It is one thing if the pilot works for a big airline like Delta or Southwest Airlines, and they are well-trained in procedure. They have little reason to miss filing a bird strike report other than the time it takes. But an independent pilot who owns his own small aircraft may be slightly more of a wildcard. If the damage is minor or nonexistant, perhaps he just shrugs it off and does not report it. If he hits a protected species like the bald eagle, he may be even less inclined as he is unsure of the consequences. 

Self-reporting always carries a bias with it, because not everybody is going to self-report. This could skew the results in ways that do not reflect reality. For example, if the data shows large airlines are far more prone to bird strikes than general aviation aircraft, that could be due to airlines being better at reporting, not because birds collide more with airliners. 

In summary, do not get caught up in just what the data says. Also ask where it came from. What could possible bias it? Framing these questions will help frame more intelligent conclusions. 

## EXERCISE

Your boss wants to see a geographic map of all birdstrikes in the year 2023. Filter down this dataframe to pull in the `LATITUDE`, `LONGITUDE`, and the records for that `INCIDENT_YEAR` 2023. 

In [15]:
# put you code here 



### SCROLL DOWN FOR ANSWER
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
|<br>
v 

First filter for the `INCIDENT_YEAR` being 2023, and then grab the `LATITUDE` and `LONGITUDE` columns on that resulting `DataFrame`. 

In [16]:
df[df["INCIDENT_YEAR"] == 2023][["LATITUDE","LONGITUDE"]]

LATITUDE  LONGITUDE
INDEX_NR                      
1353124    41.61627  -87.41279
1364684    61.17432 -149.99619
1365179   55.530193  13.371639
1368877    38.17439    -85.736
1369140    38.17439    -85.736
...             ...        ...
1515811    42.15991  -76.89144
1515812    42.15991  -76.89144
1516015         NaN        NaN
1516016         NaN        NaN
1516442      15.119  145.72936

[19612 rows x 2 columns]